### PR TITLE
WT-1930 Add configuration option to disable sweep of old handles

### PR DESCRIPTION
**Pull request by daveh86**

### CODE REVIEW

**Comment by daveh86** [comment]
@agorrod, would you be able to have a look at this?

File: dist/api_data.py [hunk 1]
```diff
@@ -402,7 +402,8 @@ def __cmp__(self, other):
             to close''', min=0),
         Config('close_idle_time', '30', r'''
             amount of time in seconds a file handle needs to be idle
-            before attempting to close it''', min=1, max=100000),
+            before attempting to close it. A setting of 0 will mean that older
+            handles are not closed''', min=0, max=100000),
```
Comment (agorrod) [inline dist/api_data.py]: replace ```will mean``` with ```means```. Replace ```older``` with ```idle```.

**Comment by agorrod** [comment]
@daveh86 I'm happy to talk about the implementation once you have made the documentation changes.

**Comment by daveh86** [comment]
@agorrod , mind having a look at the latest changes?

File: test/suite/test_sweep03.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# test_sweep01.py
```
Comment (agorrod) [inline test/suite/test_sweep03.py]: Should be sweep03

File: test/suite/test_sweep03.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+# Test lots of tables, number of open files and sweeping.  Run both
```
Comment (agorrod) [inline test/suite/test_sweep03.py]: This description should be updated to reflect the test.

File: test/suite/test_sweep03.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+# with and without checkpoints.
+#
+
+import fnmatch, os, shutil, run, time
+from suite_subprocess import suite_subprocess
+from wiredtiger import wiredtiger_open, stat
+from wtscenario import multiply_scenarios, number_scenarios, prune_scenarios
+import wttest
+
+class test_sweep01(wttest.WiredTigerTestCase, suite_subprocess):
```
Comment (agorrod) [inline test/suite/test_sweep03.py]: Should be sweep03, not sweep01

File: test/suite/test_sweep03.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        ('fix', dict(tabletype='fix',
+                    create_params = 'key_format=r,value_format=8t')),
+    ]
+
+    scenarios = types
+
+    # Overrides WiredTigerTestCase
+    def setUpConnectionOpen(self, dir):
+        self.home = dir
+        self.backup_dir = os.path.join(self.home, "WT_BACKUP")
```
Comment (agorrod) [inline test/suite/test_sweep03.py]: Is this necessary?

**Comment by agorrod** [comment]
@daveh86 The code looks OK to me. I think the test could use some work - it would be good to strip out functionality that isn't necessary for the particular case you are targeting (checkpoints, for example).

There is one other thing I'd like to test, that dropped handles are being cleaned up. To test that I'd do:

* Create a table
* Add a lot of data to that table.
* Get statistics for cache usage.
* Drop the table.
* Wait until sweep has done at least one pass
* Get statistics for cache usage.

The second statistics should show a reduction in cache usage.

**Comment by agorrod** [comment]
@daveh86 There is a review comment outstanding, and I'd like to add the cache usage checking I described in the last comment to the test.

I'd like to do the cache checking for both forced and non-forced drops.

File: src/conn/conn_sweep.c [hunk 6]
```diff
@@ -139,7 +140,8 @@ __sweep_expire(WT_SESSION_IMPL *session)
 		    F_ISSET(dhandle, WT_DHANDLE_DEAD))
 			continue;
 		if (dhandle->session_inuse != 0 ||
-		    now <= dhandle->timeofdeath + conn->sweep_idle_time)
+		    now <= dhandle->timeofdeath + conn->sweep_idle_time ||
+		    conn->sweep_idle_time == 0)
```
Comment (agorrod) [inline src/conn/conn_sweep.c]: I'd move this test to early in ```__sweep_expire```. The function is a no-op if we aren't expiring handles.

I'd change this to be:

```
/* Never expire handles if idle timeout is disabled. */
if (conn->sweep_idle_time == 0)
    return (0);
```

Just before the call to ```__wt_seconds```.

File: src/conn/conn_sweep.c [hunk 7]
```diff
@@ -266,7 +268,8 @@ __sweep_server(void *arg)
 		WT_ERR(__sweep_mark(session, &dead_handles));
 
 		if (dead_handles == 0 &&
-		    conn->open_file_count < conn->sweep_handles_min)
+		    conn->open_file_count < conn->sweep_handles_min &&
```
Comment (agorrod) [inline src/conn/conn_sweep.c]: I think this should be:

```
if (dead_handles == 0 &&
    (conn->open_file_count < conn->sweep_handles_min ||
    conn->sweep_idle_time != 0))
```

i.e: We never want to skip closing handles if we see any dead handles.
We want to skip closing handles if either there aren't enough open, or we've disabled closing idle handles.

The test case should be expanded to capture this:
Setting ```sweep_handles_min``` relatively low, opening and closing lots of handles, and ensuring they never get closed if ```sweep_idle_time``` is set to 0.

File: src/conn/conn_sweep.c [hunk 8]
```diff
@@ -121,6 +122,9 @@ __sweep_expire(WT_SESSION_IMPL *session)
 
 	conn = S2C(session);
 
+	if (conn->sweep_idle_time == 0)
```
Comment (agorrod) [inline src/conn/conn_sweep.c]: Add a comment - it's not obvious why an idle time of 0 means there is no work to do.

File: src/conn/conn_sweep.c [hunk 9]
```diff
@@ -266,7 +270,8 @@ __sweep_server(void *arg)
 		WT_ERR(__sweep_mark(session, &dead_handles));
 
 		if (dead_handles == 0 &&
-		    conn->open_file_count < conn->sweep_handles_min)
+		    (conn->open_file_count < conn->sweep_handles_min ||
```
Comment (agorrod) [inline src/conn/conn_sweep.c]: Add a comment explaining the conditions in this statement. It's subtle enough that I had to think quite hard about it.

**Comment by agorrod** [comment]
@daveh86 Did the latest version of the test fail before the last change to the ```__sweep_server```? I'd like to be confident that different combinations of configuration options work reliably via Python test suite.

**Comment by agorrod** [comment]
Thanks @daveh86 I made some changes to the Python test case. This lgtm now, I'll merge.

**Pull request merged**